source-mongodb/change_stream.go

```diff
@@ -0,0 +1,404 @@
+func (c *capture) streamForever(
+	ctx context.Context,
+	streams []changeStream,
+) error {
+	c.startStreamLogger()
+	defer c.stopStreamLogger()
+
+	log.Info("streaming change events indefinitely")
+
+	group, groupCtx := errgroup.WithContext(ctx)
+
+	for _, s := range streams {
+		s := s
+		group.Go(func() error {
+			defer s.ms.Close(groupCtx)
+
+			for {
+				if err := c.tryStream(groupCtx, s, nil); err != nil {
+					return fmt.Errorf("change stream for %q: %w", s.db, err)
+				}
+			}
+		})
+	}
+
+	return group.Wait()
+}
+
+// serverHello is used to get information regarding the most recent operation times from the server
+// to use as a high watermark for catching up change streams.
+type serverHello struct {
+	LastWrite lastWrite `bson:"lastWrite"`
+}
+
+type lastWrite struct {
+	OpTime helloTs `bson:"opTime"`
+}
+
+type helloTs struct {
+	Ts primitive.Timestamp `bson:"ts"`
+}
+
+// streamCatchup is similar to streamForever, but returns when `TryNext` yields no documents, or
+// when we are sufficiently caught up on reading the change stream as determined by comparing the
+// cluster time of the received event to the last majority-committed operation time of the cluster
+// at the start of catching up the stream.
+func (c *capture) streamCatchup(
+	ctx context.Context,
+	streams []changeStream,
+) error {
+	c.startStreamLogger()
+	defer c.stopStreamLogger()
+
+	// First get the current majority-committed operation time, which will be our high watermark for
+	// catching up the stream. Generally we will reach the "end" of the stream and TryNext will
+	// return `false`, but in cases where the stream has a high rate of changes we could end up
+	// tailing it forever and need some point to know that we are sufficiently caught up to stop.
+
+	// Note: Although we are using the "admin" database, this command works for any database, even
+	// if it doesn't exist, no matter what permissions the connecting user has.
+	result := c.client.Database("admin").RunCommand(ctx, bson.M{"hello": "1"})
+	var helloRes serverHello
+	if err := result.Decode(&helloRes); err != nil {
+		return fmt.Errorf("decoding server 'hello' response: %w", err)
+	}
+
+	ts := time.Now()
+	opTime := helloRes.LastWrite.OpTime.Ts
+	log.WithField("lastWriteOpTime", opTime).Info("catching up streams")
+	defer func() { log.WithField("took", time.Since(ts).String()).Info("finished catching up streams") }()
+
+	group, groupCtx := errgroup.WithContext(ctx)
+
+	for _, s := range streams {
+		s := s
+		group.Go(func() error {
+			ts := time.Now()
+			logEntry := log.WithField("database", s.db)
+			logEntry.Info("started catching up stream for database")
+			defer func() {
+				logEntry.WithField("took", time.Since(ts).String()).Info("finished catching up stream for database")
+			}()
+
+			if err := c.tryStream(groupCtx, s, &opTime); err != nil {
+				return fmt.Errorf("catching up stream for %q: %w", s.db, err)
+			} else if err := s.ms.Close(groupCtx); err != nil {
```

Comment:
Yeah probably not a bad idea. I do want to explicitly do the `Close` here and check for an error since we will cycle through this over and over again while backfilling, but having it as a `defer` also is probably not a bad idea since `Close` is idempotent and the symmetry is nice.